pkg/ring/batch.go

```diff
@@ -118,15 +134,28 @@ func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
 	// The use of atomic increments here guarantees only a single sendSamples
 	// goroutine will write to either channel.
 	for i := range sampleTrackers {
+		verifyIfShouldReturn := func() {
+			if sampleTrackers[i].remaining.Dec() == 0 {
+				if b.rpcsFailed.Inc() == 1 {
+					b.err <- sampleTrackers[i].err.Load()
+				}
+			}
+		}
+
 		if err != nil {
-			if sampleTrackers[i].failed.Inc() <= int32(sampleTrackers[i].maxFailures) {
+			errCount := sampleTrackers[i].recordError(err)
+
+			if errCount <= int32(sampleTrackers[i].maxFailures) {
+				verifyIfShouldReturn()
 				continue
 			}
+
 			if b.rpcsFailed.Inc() == 1 {
 				b.err <- err
 			}
 		} else {
 			if sampleTrackers[i].succeeded.Inc() != int32(sampleTrackers[i].minSuccess) {
+				verifyIfShouldReturn()
```

Comment:
This can call `b.rpcsFailed.Inc()`.  Why are we incrementing the `Failed` number on the no-error branch?

Reply:
We only do that when there is no remaining ingesters to call anymore - So even though the last ingester succeeded, we need to return an error. (in other words, because there is cases where we return an error even though the ingesters call succeeded).

lets consider the following scenario:

First Request:  4xx
Second Request: 5xx (no quorum yet)
Third Request: 2xx (in this case we did not reach maxFailures for 4xx nor 5xx)

In this particular scenario we can see that even though we have 2 errors (4xx + 5xx) we did not returned any error yet (as we had no quorum). What the `verifyIfShouldReturn` is doing is basically checking that there is no remaining calls to do for this particular sample and id not, return an error (pick any of the 4xx or 5xx). Its important to note that this function is called only if we did not reach the `minSuccess` needed to return a success.

Ps:
`b.rpcsFailed.Inc()` is here only to make sure we only return one error for the whole batch.